modyn/selector/internal/selector_strategies/presampling_strategies/random_no_replacement_presampling_strategy.py

```diff
@@ -0,0 +1,94 @@
+from typing import Optional
+
+from modyn.metadata_database.metadata_database_connection import MetadataDatabaseConnection
+from modyn.metadata_database.models import SelectorStateMetadata
+from modyn.selector.internal.selector_strategies.presampling_strategies.abstract_presampling_strategy import (
+    AbstractPresamplingStrategy,
+)
+from sqlalchemy import Select, asc, func, select
+
+
+class RandomNoReplacementPresamplingStrategy(AbstractPresamplingStrategy):
+    def __init__(self, presampling_config: dict, modyn_config: dict, pipeline_id: int):
+        super().__init__(presampling_config, modyn_config, pipeline_id)
+        self.requires_trigger_dataset_size = True
+
+        # a complete_trigger is the last time when all the datapoints (in the db at that time) have been seen
+        self.last_complete_trigger = 0
+
+    def get_presampling_query(
+        self,
+        next_trigger_id: int,
+        tail_triggers: Optional[int],
+        limit: Optional[int],
+        trigger_dataset_size: Optional[int],
+    ) -> Select:
+        assert trigger_dataset_size is not None
+        assert trigger_dataset_size >= 0
+
+        target_size = self.get_target_size(trigger_dataset_size, limit)
+
+        # get the query to sample up to target size points that have not been used since last_complete_trigger
+        subq = self._get_query_sample_random(next_trigger_id, tail_triggers, target_size)
+
+        # update last_used_in_trigger of the samples selected in the subquery
+        self._update_last_used_in_trigger(next_trigger_id, subq)
```

Comment:
Should we move this to the CoresetStrategy level instead of doing this in every presampling strategy? Or is there a reason that needs to be per persampling strategy?

Reply:
Right now, only randomNoReplacement writes to the database to select samples. The other strategies do not use it, I think it would be much more inefficient otherwise

Reply:
ahh, yes, I somehow thought all strategies use it. then it makes sense to just have that here.